cmd/xpumanager_sidecar/main.go

```diff
@@ -0,0 +1,355 @@
+
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/intel/intel-device-plugins-for-kubernetes/cmd/internal/pluginutils"
+	"k8s.io/klog/v2"
+
+	io_prometheus_client "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+)
+
+type invalidEntryErr struct{}
+
+const (
+	labelMaxLength        = 63
+	xeLinkLabelName       = "xe-links"
+	pureXeLinkMetricValue = 1
+)
+
+type xpuManagerTopologyMatrixCell struct {
+	LocalDeviceID     int
+	LocalSubdeviceID  int
+	RemoteDeviceID    int
+	RemoteSubdeviceID int
+	LaneCount         int
+}
+
+type xpuManagerSidecar struct {
+	getMetricsData func() []byte
+	tmpDirPath     string
+	dstFilePath    string
+	labelNamespace string
+	url            string
+	interval       uint64
+	startDelay     uint64
+	xpumPort       uint64
+	laneCount      uint64
+}
+
+func (e *invalidEntryErr) Error() string {
+	return "metrics entry was invalid for our use"
+}
+
+func (xms *xpuManagerSidecar) getMetricsDataFromXPUM() []byte {
+	client := &http.Client{
+		Timeout: 5 * time.Second,
+	}
+
+	ctx := context.Background()
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, xms.url, http.NoBody)
+	if err != nil {
+		klog.Error(err.Error())
+
+		return nil
+	}
+
+	res, err := client.Do(req)
+	if err != nil {
+		klog.Error(err.Error())
+
+		return nil
+	}
+
+	resBody, err := io.ReadAll(res.Body)
+
+	defer res.Body.Close()
+
+	if err != nil {
+		klog.Error(err.Error())
+
+		return nil
+	}
+
+	// Seems /metrics doesn't add new-line at the end of the last entry
+	// and without this there is an error from TextParser
+	resBody = append(resBody, "\n"...)
+
+	return resBody
+}
+
+func processMetricsLabels(labels []*io_prometheus_client.LabelPair) (xpuManagerTopologyMatrixCell, error) {
+	cell := xpuManagerTopologyMatrixCell{}
+	keepGoing := true
+
+	// Atoi shouldn't ever fail. TextParser fails lines with invalid value
+	getValueFromString := func(stringValue string, out *int) {
+		val, err := strconv.Atoi(stringValue)
+		if err == nil {
+			*out = val
+		} else {
+			keepGoing = false
+		}
+	}
+
+	lanCountFound := false
+
+	for _, label := range labels {
+		klog.V(5).Info(label.GetName(), " ", label.GetValue())
+
+		switch name := label.GetName(); name {
+		case "local_device_id":
+			getValueFromString(label.GetValue(), &cell.LocalDeviceID)
+		case "local_subdevice_id":
+			getValueFromString(label.GetValue(), &cell.LocalSubdeviceID)
+		case "remote_device_id":
+			getValueFromString(label.GetValue(), &cell.RemoteDeviceID)
+		case "remote_subdevice_id":
+			getValueFromString(label.GetValue(), &cell.RemoteSubdeviceID)
+		case "lan_count":
+			getValueFromString(label.GetValue(), &cell.LaneCount)
+
+			lanCountFound = true
+		case "local_on_subdevice":
+			// xelinks should always be on subdevices
+			if label.GetValue() != "true" {
+				keepGoing = false
+			}
+		}
+
+		if !keepGoing {
```

Comment:
IMHO this would be clearer if you would drop `getValueFromString()` func & `keepGoing` var, and did the loop like this:
```
name := label.GetName()
valstr := label.GetValue()

klog.V(5).Info(name, "", valstr)

// xelinks should always be on subdevices
if name == "local_on_subdevice" {
  if valstr != "true" {
     return cell, &invalidEntryErr{}
}

val, err := strconv.Atoi(valstr)
if err != nil {
   return cell, &invalidEntryErr{}
}

switch name {
case "local_device_id":
	cell.LocalDeviceID = val
```

Reply:
There are also other labels that do not have integers as values. So we can't threat any integer conversion failure as a complete failure.